x/wasm/keeper/addresses_test.go

```diff
@@ -3,20 +3,82 @@ package keeper
 import (
 	"encoding/json"
 	"fmt"
+	tmbytes "github.com/cometbft/cometbft/libs/bytes"
 	"testing"
 
-	tmbytes "github.com/cometbft/cometbft/libs/bytes"
 	"github.com/stretchr/testify/require"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-func TestBuildContractAddress(t *testing.T) {
+func TestBuildContractAddressClassic(t *testing.T) {
+	// preserve current Bech32 settings and restore them after test completion
 	x, y := sdk.GetConfig().GetBech32AccountAddrPrefix(), sdk.GetConfig().GetBech32AccountPubPrefix()
+	c := sdk.IsAddrCacheEnabled()
 	t.Cleanup(func() {
 		sdk.GetConfig().SetBech32PrefixForAccount(x, y)
+		sdk.SetAddrCacheEnabled(c)
 	})
+
+	// set custom Bech32 settings
 	sdk.GetConfig().SetBech32PrefixForAccount("purple", "purple")
+	// disable address cache
+	// AccAddress -> String conversion is then slower, but does not lead to errors like this:
+	//   runtime error: invalid memory address or nil pointer dereference
+	sdk.SetAddrCacheEnabled(false)
+
+	// run tests
+	for i, spec := range classicContractAddr {
+		t.Run(fmt.Sprintf("case %d", i), func(t *testing.T) {
+			// when
+			gotAddr := BuildContractAddressClassic(spec.codeId, spec.instanceId)
+			// then
+			require.Equal(t, spec.address, gotAddr.String())
+			require.NoError(t, sdk.VerifyAddressFormat(gotAddr))
+		})
+	}
+}
+
+var classicContractAddr = []struct {
+	codeId     uint64
+	instanceId uint64
+	address    string
+}{
+	{
+		codeId:     0,
+		instanceId: 0,
+		address:    "purple1w0w8sasnut0jx0vvsnvlc8nayq0q2ej8xgrpwgel05tn6wy4r57qfplul7",
+	},
+	{
+		codeId:     0,
+		instanceId: 1,
+		address:    "purple156r47kpk4va938pmtpuee4fh77847gqcw2dmpl2nnpwztwfgz04s5cr8hj",
+	},
+	{
+		codeId:     1,
+		instanceId: 0,
+		address:    "purple1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfs5efvjk",
+	},
+	{
+		codeId:     1,
+		instanceId: 1,
+		address:    "purple14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9smc2vxm",
+	},
+}
```

Comment:
you could declare and initialize this directly inside the test function because it is closely related to it and definitely we are not using somewhere else.
Moreover, as personal preference I would call `expAddress` instead of `address`, just to make it clear what are the inputs and what are the expected values.

Reply:
Done.